### PR TITLE
Windows testing support for Rasa

### PR DESCRIPTION
**Pull request by alwx**
**Proposed changes**:
- ...

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: tests/utilities.py [hunk 1]
```diff
@@ -1,14 +1,28 @@
...
+from yarl import URL
+from typing import List, Text
+
 from rasa.nlu.classifiers.diet_classifier import DIETClassifier
 from rasa.nlu.classifiers.embedding_intent_classifier import EmbeddingIntentClassifier
 from rasa.nlu.selectors.response_selector import ResponseSelector
 from rasa.utils.tensorflow.constants import EPOCHS
 
 
+def write_temp_file(contents, suffix, mode="w+") -> Text:
```
Comment (ricwo) [inline tests/utilities.py:14]:
```suggestion
def write_temp_file(contents: AnyStr, suffix: Text, mode: Optional[Text] = "w+") -> Text:
```

File: tests/utilities.py [hunk 2]
```diff
@@ -1,14 +1,28 @@
...
+    filename = os.path.join(tempfile.gettempdir(), os.urandom(24).hex() + suffix)
```
Comment (ricwo) [inline tests/utilities.py:15]: optional: use pathlib.Path

File: tests/utilities.py [hunk 3]
```diff
@@ -17,6 +31,10 @@ def json_of_latest_request(r):
     return r[-1].kwargs["json"]
 
 
+def platform_independent_paths(coll: List[Text]):
```
Comment (ricwo) [inline tests/utilities.py:34]:
```suggestion
def platform_independent_paths(coll: List[Text]) -> List[Text]:
```

File: tests/test_server.py [hunk 4]
```diff
@@ -689,7 +690,7 @@ def test_pushing_event(rasa_app: SanicTestClient, event: Event):
     evt = tracker.get("events")[0]
     deserialised_event = Event.from_parameters(evt)
     assert deserialised_event == event
-    assert deserialised_event.timestamp > time_before_adding_events
+    assert deserialised_event.timestamp >= time_before_adding_events
```
Comment (ricwo) [inline tests/test_server.py:693]: why this change?

File: tests/utilities.py (hunk 1 above)
Comment (ricwo) [inline tests/utilities.py:14]: `utils.io.create_temporary_file()` does almost the same thing. can we use that instead?

File: tests/nlu/test_utils.py [hunk 5]
```diff
@@ -32,7 +32,10 @@ def fake_model_dir(empty_model_dir):
 
 def test_relative_normpath():
     test_file = "/my/test/path/file.txt"
-    assert utils.relative_normpath(test_file, "/my/test") == "path/file.txt"
+    assert (
+        utils.relative_normpath(test_file, "/my/test").replace("\\", "/")
+        == "path/file.txt"
+    )
```
Comment (ricwo) [inline tests/nlu/test_utils.py:38]: this should be platform-agnostic without the `replace()`
```suggestion
    assert (
        assert Path(utils.relative_normpath(test_file, "/my/test")) == Path("path/file.txt")
    )

File: tests/nlu/test_utils.py [hunk 6]
```diff
@@ -32,7 +32,10 @@ def fake_model_dir(empty_model_dir):
 
 def test_relative_normpath():
```
Comment (ricwo) [inline tests/nlu/test_utils.py:33]: we could also update the utils function to use `Path(p).resolve()` - it replaces `os.path.normpath()` as of py 3.6: https://docs.python.org/3.6/library/pathlib.html#pathlib.Path.resolve

File: tests/core/test_data.py [hunk 7]
```diff
@@ -157,8 +158,8 @@ def test_same_file_names_get_resolved(tmpdir):
 def test_find_nlu_files_with_different_formats(test_input, expected):
     examples_dir = "data/examples"
     data_dir = os.path.join(examples_dir, test_input)
-    core_files, nlu_files = data.get_core_nlu_files([data_dir])
-    assert nlu_files == expected
+    paths = data.get_core_nlu_files([data_dir])
+    assert platform_independent_paths(paths[1]) == expected
```
Comment (ricwo) [inline tests/core/test_data.py:162]:
```suggestion
    assert Path(paths[1]) == Path(expected)
```

File: Makefile [hunk 8]
```diff
@@ -70,16 +81,26 @@ prepare-tests-files:
...
 	brew install graphviz || true
 
 prepare-tests-ubuntu: prepare-tests-files
 	sudo apt-get -y install graphviz graphviz-dev python-tk
 
+prepare-tests-windows: prepare-wget-windows prepare-tests-files
+	choco install graphviz
+
 test: clean
 	# OMP_NUM_THREADS can improve overral performance using one thread by process (on tensorflow), avoiding overload
```
Comment (ricwo) [inline Makefile:97]:
```suggestion
	# OMP_NUM_THREADS can improve overal performance using one thread by process (on tensorflow), avoiding overload
```

File: Makefile [hunk 9]
```diff
@@ -70,16 +81,26 @@ prepare-tests-files:
...
 	OMP_NUM_THREADS=1 poetry run pytest tests -n $(JOBS) --cov rasa
 
+test-windows: clean
+	# OMP_NUM_THREADS can improve overral performance using one thread by process (on tensorflow), avoiding overload
```
Comment (ricwo) [inline Makefile:101]:
```suggestion
	# OMP_NUM_THREADS can improve overal performance using one thread by process (on tensorflow), avoiding overload
```

File: tests/utilities.py [hunk 10]
```diff
@@ -17,6 +31,10 @@ def json_of_latest_request(r):
...
+    return [i.replace("\\", "/") for i in coll]
```
[the remark this answers is not in the document]
Comment (ricwo) [reply tests/utilities.py:35]: in that case we could get rid of this function

File: Makefile (hunk 8 above)
Comment (wochinge) [reply Makefile:97]: I changed it to `overall` (two `l`s)

File: tests/core/test_lock_store.py [hunk 11]
```diff
@@ -218,6 +218,7 @@ async def mocked_handle_message(self, message: UserMessage, wait: float) -> None
         assert time.time() - start_time < time_limit
 
 
+@pytest.mark.unix
```
Comment (wochinge) [inline tests/core/test_lock_store.py:221]: what's the reason here?

File: tests/core/test_tracker_stores.py [hunk 12]
```diff
@@ -45,6 +44,7 @@
 from rasa.core.trackers import DialogueStateTracker
 from rasa.utils.endpoints import EndpointConfig, read_endpoint_config
 from tests.core.conftest import DEFAULT_ENDPOINTS_FILE, MockedMongoTrackerStore
+from tests.utilities import write_file_config
```
Comment (wochinge) [inline tests/core/test_tracker_stores.py:47]: please import the module instead of the function

File: tests/core/test_tracker_stores.py [hunk 13]
```diff
@@ -559,12 +556,12 @@ def test_tracker_store_retrieve_with_session_started_events(
 ):
     tracker_store = tracker_store_type(default_domain, **tracker_store_kwargs)
     events = [
-        UserUttered("Hola", {"name": "greet"}),
-        BotUttered("Hi"),
-        SessionStarted(),
-        UserUttered("Ciao", {"name": "greet"}),
+        UserUttered("Hola", {"name": "greet"}, timestamp=1),
```
Comment (wochinge) [inline tests/core/test_tracker_stores.py:559]: do we need this because the timestamps on Windows would be the same?

File: tests/test_server.py [hunk 14]
```diff
@@ -207,6 +207,7 @@ def send_request() -> None:
     train_request.terminate()
 
 
+@pytest.mark.unix
```
Comment (wochinge) [inline tests/test_server.py:210]: why is this one required?

File: tests/utilities.py [hunk 15]
```diff
@@ -1,14 +1,28 @@
-from yarl import URL
-from typing import Text
-
+import tempfile
+import os
 import rasa.utils.io as io_utils
 
+from yarl import URL
```
Comment (wochinge) [inline tests/utilities.py:5]: this is either the wrong import or it should be added to the dev dependencies

File: rasa/cli/scaffold.py [hunk 16]
```diff
@@ -42,11 +42,12 @@ def print_train_or_instructions(args: argparse.Namespace, path: Text) -> None:
...
 
-    should_train = (
-        questionary.confirm("Do you want to train an initial model? 💪🏽")
-        .skip_if(args.no_prompt, default=True)
-        .ask()
-    )
+    if args.no_prompt:
+        should_train = True
+    else:
+        should_train = questionary.confirm(
```
[the remark this answers is not in the document]
Comment (alwx) [reply rasa/cli/scaffold.py:48]: It just doesn't work on Windows :( (it's an issue with `questionary`)

File: tests/cli/test_rasa_init.py [hunk 17]
```diff
@@ -30,13 +30,10 @@ def test_init_using_init_dir_option(run: Callable[..., RunResult]):
...
 
-def test_not_fount_init_path(run: Callable[..., RunResult]):
+def test_not_found_init_path(run: Callable[..., RunResult]):
     output = run("init", "--no-prompt", "--quiet", "--init-dir", "./workspace")
 
-    assert (
-        output.outlines[-1]
-        == "\033[91mProject init path './workspace' not found.\033[0m"
-    )
+    assert "Project init path './workspace' not found." in output.outlines[-1]
```
[the remark this answers is not in the document]
Comment (alwx) [reply tests/cli/test_rasa_init.py:36]: Have absolutely no idea why, but this test returns text without any color codes on Windows, just plain text.

File: tests/utilities.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (alwx) [reply tests/utilities.py:35]: 👍

File: tests/core/test_tracker_stores.py (hunk 13 above)
Comment (alwx) [reply tests/core/test_tracker_stores.py:559]: yes, all the timestamps on windows are the same, and the order is not consistent

File: rasa/cli/scaffold.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/cli/scaffold.py:48]: @alwx can we fix that in the questionary package? do you know what the underlying issue is?

[the remark this answers is not in the document]
Comment (alwx) [reply rasa/cli/scaffold.py:48]: Had no chance to check it yet. Will try to find time for it tomorrow, maybe it's an easy fix.
